### PR TITLE
[SPARK-35020][SQL] Group exception messages in catalyst/util

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR group exception messages in `sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util`.

### Why are the changes needed?
It will largely help with standardization of error messages and its maintenance.

### Does this PR introduce _any_ user-facing change?
No. Error messages remain unchanged.

### How was this patch tested?
No new tests - pass all original tests to make sure it doesn't break any existing behavior.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @allisonwang-db

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 1]
```diff
@@ -823,4 +824,55 @@ object QueryExecutionErrors {
...
+  }
+
+  def duplicateMapKeyFoundError(key: Any): Throwable = {
+    new RuntimeException(s"Duplicate map key $key was found, please check the input " +
+      "data. If you want to remove the duplicated keys, you can set " +
+      s"${SQLConf.MAP_KEY_DEDUP_POLICY.key} to ${SQLConf.MapKeyDedupPolicy.LAST_WIN} so that " +
+      "the key inserted at last takes precedence.")
+  }
+
+  def differentLengthOfKeyArrayAndValueArrayOfMapDataError(): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:840]: mapDataKeyArrayLengthDiffersFromValueArrayLengthError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 2]
```diff
@@ -823,4 +824,55 @@ object QueryExecutionErrors {
...
+    new RuntimeException("The key array and value array of MapData must have the same length.")
+  }
+
+  def fieldDiffersFromDerivedOneError(
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:844]: maybe `fieldDiffersFromDerivedLocalDateError`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 3]
```diff
@@ -823,4 +824,55 @@ object QueryExecutionErrors {
...
+      field: ChronoField, actual: Int, expected: Int, candidate: LocalDate): Throwable = {
+    new DateTimeException(s"Conflict found: Field $field $actual differs from" +
+      s" $field $expected derived from $candidate")
+  }
+
+  def failedParseDateTimeInNewParserError(s: String, e: Throwable): Throwable = {
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:850]: failToParseDateTimeInNewParserError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 4]
```diff
@@ -823,4 +824,55 @@ object QueryExecutionErrors {
...
+    new SparkUpgradeException("3.0", s"Fail to parse '$s' in the new parser. You can " +
+      s"set ${SQLConf.LEGACY_TIME_PARSER_POLICY.key} to LEGACY to restore the behavior " +
+      s"before Spark 3.0, or set to CORRECTED and treat it as an invalid datetime string.", e)
+  }
+
+  def failedFormatDateTimeInNewFormatterError(
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:856]: failToFormatDateTimeInNewFormatterError

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 5]
```diff
@@ -823,4 +824,55 @@ object QueryExecutionErrors {
...
+      resultCandidate: String, e: Throwable): Throwable = {
+    new SparkUpgradeException("3.0",
+      s"""
+         |Fail to format it to '$resultCandidate' in the new formatter. You can set
+         |${SQLConf.LEGACY_TIME_PARSER_POLICY.key} to LEGACY to restore the behavior before
+         |Spark 3.0, or set to CORRECTED and treat it as an invalid datetime string.
+       """.stripMargin.replaceAll("\n", " "), e)
+  }
+
+  def failedRecognizePatternInDateTimeFormatterError(
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:866]: failToRecognizePatternInDateTimeFormatterError

**Review by allisonwang-db — COMMENTED** [review]
Looks good!

**Comment by beliefer** [comment]
ping @allisonwang-db

**Review by allisonwang-db — APPROVED** [review]
Looks good!

**Comment by beliefer** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@allisonwang-db Thank you for review. @cloud-fan Thank you too.

**Pull request closed without merging**